src/main/java/org/sert2521/gamename/Log.kt

```diff
@@ -0,0 +1,34 @@
+package org.sert2521.gamename
+
+import edu.wpi.first.wpilibj.Preferences
+
+/** Checks if a log is loggable with the provided log level from SmartDashboard. */
+private fun isLoggable(level: LogLevel) =
+        level >= when (Preferences.getInstance().getString("log_level", "info")) {
+            "verbose" -> LogLevel.VERBOSE
+            "debug" -> LogLevel.DEBUG
+            "warn" -> LogLevel.WARN
+            "error" -> LogLevel.ERROR
+            else -> LogLevel.INFO
+        }
+
+enum class LogLevel { VERBOSE, DEBUG, INFO, WARN, ERROR }
+
+private val autoTag: String get() {
+    Thread.currentThread().stackTrace
+            .filter { it.fileName != "LogKt" }
+            .forEach { return it.className }
+    return ""
+}
+
+fun verbose(tag: String? = null, message: Any) = log(LogLevel.VERBOSE, tag, message)
+fun debug(tag: String? = null, message: Any) = log(LogLevel.DEBUG, tag, message)
+fun info(tag: String? = null, message: Any) = log(LogLevel.INFO, tag, message)
+fun warn(tag: String? = null, message: Any) = log(LogLevel.WARN, tag, message)
+fun error(tag: String? = null, message: Any) = log(LogLevel.ERROR, tag, message)
```

Comment:
@andrewda Yeah so Anko is using pretty much the most advanced stuff you'll find in Kotlin: extension and inline functions. Once you get it, it's not too hard so I'll try to explain it.

Here's their method:
```kotlin
fun AnkoLogger.debug(...) {
    log(this, message, ...)
}
```

Let's break this apart. `AnkoLogger.debug` is saying we're an extension function meaning we're extending an existing class. This is how Kotlin can provide methods like [`filter`](https://github.com/SouthEugeneRoboticsTeam/Robot-Base/pull/2/files#diff-bbae703ef686b19f885a7f38757aa186R19) which don't exist in the Java version of `List`. So basically `AnkoLogger.debug` is saying we're now inside this class (except you can only access public members because it's compiled down to a static method with the instance as a parameter in Java).

`log(this, ...)` we can do that because remember we are _inside_ `AnkoLogger`.

Now:
```kotlin
private inline fun log(
        logger: AnkoLogger,
        message: Any?,
        thr: Throwable?,
        level: Int,
        f: (String, String) -> Unit,
        fThrowable: (String, String, Throwable) -> Unit) {
    val tag = logger.loggerTag
    if (Log.isLoggable(tag, level)) {
        if (thr != null) {
            fThrowable(tag, message?.toString() ?: "null", thr)
        } else {
            f(tag, message?.toString() ?: "null")
        }
    }
}
```

Here, the `inline` keyword tells the Kotlin compiler to move the code inside the inline function to where it's used to it doesn't have to create an anonymous inner class which is expensive if you're calling it a lot of times. Then `val tag = logger.loggerTag` is getting the class name from that interface you saw. Remember that we were passing in `this` in the previous method which is the class you are inside e.g. the one implementing the interface.

So basically, by implementing that interface, Anko can move itself into your class to pull out its name. The one thing that annoys me with the way they did it is that you actually have to implement their interface which is 💩.

This is a great video explaining these concepts from Android God Jake Wharton: https://youtu.be/A2LukgT2mKc?t=7m59s
If you're a bad person and you don't want to watch the whole video, here's where the hardcore stuff starts: https://youtu.be/A2LukgT2mKc?t=23m46s

I would **_very strongly_** recommend watching the whole thing from where I started it in the first link. You could watch if from the beginning, but it's just going over syntax which you're probably familiar with because JS.

Holly shit that was long! 😄